### PR TITLE
add client-side embeddings cached in IndexedDB

### PR DESCRIPTION
**Pull request by Mbeaulne**
## Description

Adds embedding-based semantic search as a pre-processing step before LLM reranking. When an AI provider API base is configured, the component search now fetches text embeddings for both the search query and all indexed components, ranks them by cosine similarity, and merges those results with the top lexical matches before passing candidates to the LLM reranker. This improves the quality of candidates surfaced to the reranker, particularly for queries that don't share keywords with component names or descriptions.

A new `componentSearchEmbeddings` service handles embedding generation, cosine similarity scoring, and persistent caching of embeddings in IndexedDB (keyed by a FNV-1a hash of the text and model name) to avoid redundant API calls across searches. A `mergeUniqueMatches` helper deduplicates and merges lexical and embedding results by digest before reranking.

The embedding fetch is tracked with an `isEmbeddingSearchPending` flag, which gates the rerank active state and disables the AI search buttons while in progress, keeping the spinner and disabled states consistent with the existing reranking UX.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [x] Improvement
- [ ] Cleanup/Refactor
- [ ] Breaking change
- [ ] Documentation update

## Checklist

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Configure an OpenAI-compatible API base (e.g. `https://api.openai.com/v1`) in AI provider settings.
2. Open the component search panel and enter a query that is semantically related to a component but does not share exact keywords.
3. Click the AI search (Sparkles) button and verify the spinner appears during both the embedding fetch and reranking phases.
4. Confirm that results are ranked by semantic relevance.
5. Repeat the same search and verify via network tooling that no additional embedding API calls are made (cache hit).
6. Verify that searches still work correctly when no API base is configured (embedding step is skipped gracefully).

## Additional Comments

The embedding model is hardcoded to `text-embedding-3-small`. If no `apiBase` is set, the embedding step is skipped entirely and behaviour is identical to before this change.

### CODE REVIEW

**Comment by Mbeaulne** [comment]
* **#2462** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2462?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2461** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2461?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2459** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2459?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2458** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2457** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2456** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2456?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2454** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2453** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2453?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2452** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2452?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2451** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2451?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2450** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2450?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2449** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2449?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2448** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2447** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2447?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2444** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2433** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2433?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2432** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2432?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2432?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2431** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2430** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2429** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2428** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2427** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2427?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2426** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2426?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2425** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2425?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2424** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2424?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2423** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2423?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/services/componentSearchEmbeddings.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    ) {
+      result.set(uniqueTexts[i], cached.embedding);
+    } else {
+      missingTexts.push(uniqueTexts[i]);
+    }
+  }
+
+  if (missingTexts.length === 0) return result;
+
+  const embeddings = await fetchEmbeddings(missingTexts, options);
```
Comment (Mbeaulne) [inline src/services/componentSearchEmbeddings.ts]: > 🤖 This is an AI-generated code review comment.

[MEDIUM] The full unfiltered index is embedded on every search, and `embedTextsWithCache` sends all cache-missing texts in a single fetch body with no chunking. On a cold cache for a large library this exceeds OpenAI embeddings limits (2048 inputs / ~300k tokens) → 400, and the whole semantic layer silently returns `[]` (swallowed by the host catch). Consider batching missing texts (e.g. 256–512/request) and merging, embedding only the candidate set, or capping N.

File: src/services/componentSearchEmbeddings.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  let hash = 2166136261;
+  for (let i = 0; i < text.length; i++) {
+    hash ^= text.charCodeAt(i);
+    hash = Math.imul(hash, 16777619);
+  }
+  return (hash >>> 0).toString(16);
+}
+
+function cacheKey(text: string): string {
+  return `${CACHE_SCHEMA_VERSION}:${COMPONENT_SEARCH_EMBEDDING_MODEL}:${hashText(text)}`;
```
Comment (Mbeaulne) [inline src/services/componentSearchEmbeddings.ts:59]: > 🤖 This is an AI-generated code review comment.

[LOW/MEDIUM] The cache key and the `textHash` validation field both derive from the same 32-bit FNV-1a hash, so a key collision (~50% likelihood near ~77k distinct texts) returns a stale-but-"validated" vector with no detection. Consider storing the full text (or a wider hash, e.g. SHA-256, or including text length) as the validation field so a key collision is caught as a miss.

File: src/services/componentSearchEmbeddings.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    } else {
+      missingTexts.push(uniqueTexts[i]);
+    }
+  }
+
+  if (missingTexts.length === 0) return result;
+
+  const embeddings = await fetchEmbeddings(missingTexts, options);
+  const now = Date.now();
+  await componentSearchEmbeddingDb.embeddings.bulkPut(
```
Comment (Mbeaulne) [inline src/services/componentSearchEmbeddings.ts]: > 🤖 This is an AI-generated code review comment.

[LOW] Dexie schema is pinned at `version(1)`; cache busting relies on `CACHE_SCHEMA_VERSION` embedded in the key, so bumping it orphans all prior rows permanently (no prune). Also, there's no `QuotaExceededError` handling on this `bulkPut` — the rejection is swallowed by the host catch, silently disabling caching. Consider pruning rows whose key lacks the current `schemaVersion:model:` prefix on a schema bump, and catching quota errors to trigger a prune.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 4]
```diff
@@ -437,6 +439,23 @@ function sampleEvenly<T>(entries: T[], limit: number): T[] {
...
 }
 
+function mergeUniqueMatches(
+  primary: LexicalMatch[],
+  secondary: LexicalMatch[],
+  fallback: LexicalMatch[],
+  limit: number,
+): LexicalMatch[] {
+  const merged: LexicalMatch[] = [];
+  const seen = new Set<string>();
```
Comment (Mbeaulne) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx:460]: > 🤖 This is an AI-generated code review comment.

[LOW] This `mergeUniqueMatches` (dedupe-by-digest) is duplicated verbatim in `useComponentSearchV2State.ts` (~line 53). Consider exporting it once (e.g. from `componentSearchIndex.ts`) and importing it in both hosts.

File: src/routes/Dashboard/DashboardComponentsV2View.tsx [hunk 5]
```diff
@@ -780,27 +802,57 @@ export const DashboardComponentsV2View = () => {
...
-    setRerankBaseMatches(matches);
+    setRerankBaseMatches(rerankMatches);
     setRerankedFor(trimmed);
     rerank({ query: trimmed, candidates });
   };
 
   const handleSmartSearch = () => {
-    startAiSearch(aiCandidateMatches);
+    void startAiSearch(aiCandidateMatches, aiCandidateMatches.length);
   };
```
Comment (Mbeaulne) [inline src/routes/Dashboard/DashboardComponentsV2View.tsx:857]: > 🤖 This is an AI-generated code review comment.

`maxTypoDistance` allows edit distance 1 on 4-char tokens, so short generic IO names collide: `data`<->`date`, `path`<->`bath`, `list`<->`last`. Fuzzy is name/io-only and scored at 0.75x so exact hits still win, but a typo-free `data` could pull in `date`-named IO. Consider raising the distance-1 floor to length >= 5.

**Comment by camielvs** [comment]
## 🤖 Code review — Add client-side embeddings cached in IndexedDB

Nice addition of a semantic retrieval layer. The core design is sound: a content-hash cache key namespaced by schema version + model, `bulkGet`/`bulkPut` with `Set` dedup, graceful degradation to lexical on any failure, a correct cosine impl with a zero-norm guard, and a test that proves the cache path (second identical search → 0 fetches). Hybrid merge (top-60 lexical + embedding hits + fallback) is a reasonable way to feed the reranker. A handful of things to address:

### Correctness / robustness

- **No batching of the embeddings request.** `embedTextsWithCache` sends *all* missing texts in one `/embeddings` call (`input: texts`). The first search on a large library puts every component's text in a single request — OpenAI-compatible endpoints cap inputs per request (item count and total tokens). Exceed that and the call throws → caught → embeddings silently return `[]` (no semantic benefit, wasted round-trip). Chunk `missingTexts` into batches (e.g. 96–256) and merge results.

- **The `textHash` collision guard is a no-op.** `cacheKey` is already `version:model:hashText(text)`, so any row returned by `bulkGet(keys[i])` necessarily has the same hash — and the guard then re-checks `cached.textHash === hashText(text)`, the *same* 32-bit FNV hash. So it provides zero protection against a hash collision: two distinct texts that collide return the wrong cached embedding (and thus wrong similarity). Negligible at hundreds of components, but the query-embedding cache (below) accumulates unboundedly, raising the odds over time. To actually guard, store the source text and compare it; otherwise the redundant `textHash`/`embeddingModel` fields can go.

- **Unbounded IndexedDB growth.** `updatedAt` is written but never read — there's no TTL, size cap, or eviction, and `clearComponentSearchEmbeddingCache` is only called in tests. Every unique *query* is embedded and cached forever, so the table grows across sessions without limit (each 1536-dim vector is ~10KB+ as JSON). Add an eviction policy (TTL or max-rows prune), particularly for query embeddings.

- **Embedding fetches are uncancellable.** `EmbeddingOptions` has a `signal` field, but neither caller wires one up (confirmed). Changing the query mid-flight leaves the billed embedding request running. This repo explicitly removed "uncancellable billed calls" for AI descriptions (see the `useComponentAiDescription` comment) — this reintroduces the pattern. Thread an `AbortSignal` tied to query change / unmount.

### Duplication

- **`mergeUniqueMatches` is copy-pasted verbatim** into both `DashboardComponentsV2View` and `useComponentSearchV2State`. And `buildComponentEmbeddingText` is now a *third* independent component→text projection alongside the index's `searchable` builder and `componentReferenceToCandidate` — note it doesn't apply the annotation-key exclusions or length caps the index uses, so the three can drift. Extract shared helpers; the embedding text in particular should probably reuse the index's extraction.

### Minor

- **Broader data egress.** The first embedding search sends the *entire* library's text — names, descriptions, I/O, and `published_by` emails — to the external embeddings endpoint, which is broader than the rerank (candidates only). It's opt-in via `apiBase` and consistent with existing AI features, but worth a conscious sign-off since it's now every component, not just matches.
- **Editor embeds the full `index`; Dashboard embeds `filteredIndex`.** Confirm the source-filter divergence is intentional.
- Good test coverage for the happy paths; consider adding one for a >1-batch miss and for cache invalidation when a component's text changes (new hash → re-embed only that entry).

File: src/services/componentSearchEmbeddings.ts (hunk 1 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.ts]: > 🤖 This is an AI-generated response.

Fixed (working tree, pending commit) — `embedTextsWithCache` now batches `missingTexts` into chunks of `EMBEDDING_BATCH_SIZE` (256), calling `fetchEmbeddings` per chunk and concatenating, so a large cold-cache library no longer sends the whole index in one request and 400s. Existing tests (fake-indexeddb) still pass.

File: src/services/componentSearchEmbeddings.ts (hunk 2 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.ts:59]: > 🤖 This is an AI-generated response.

Fixed (working tree, pending commit) — added a `textLength` field to the cache entry and the read-side validation now requires `cached.textHash === hashText(t) && cached.textLength === t.length`. A 32-bit-hash collision now also has to match the text length, so a colliding entry is caught as a cache miss (and re-embedded) rather than returning a stale vector. No storage bloat (an int, not the full text), and old entries without `textLength` are simply treated as misses and refreshed once.

File: src/services/componentSearchEmbeddings.ts (hunk 3 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.ts]: > 🤖 This is an AI-generated response.

Fixed the quota half (working tree, pending commit) — the `bulkPut` is now wrapped in try/catch, so a `QuotaExceededError` (or any IndexedDB write failure) no longer propagates into the host catch and silently returns `[]`; the computed embeddings are still returned, only the cache write is skipped (degrades to "recompute next time").

Deferring the prune-on-schema-bump half to a follow-up: it is more involved (delete rows whose key lacks the current `schemaVersion:model:` prefix) and `CACHE_SCHEMA_VERSION` is still 1, so no orphaning has happened yet. Tracking it separately.

File: src/services/componentSearchEmbeddings.ts [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  const now = Date.now();
+  try {
+    await componentSearchEmbeddingDb.embeddings.bulkPut(
+      missingTexts.map((text, index) => ({
+        cacheKey: cacheKey(text),
+        embeddingModel: COMPONENT_SEARCH_EMBEDDING_MODEL,
+        textHash: hashText(text),
+        textLength: text.length,
+        embedding: embeddings[index],
+        updatedAt: now,
```
Comment (camielvs) [inline src/services/componentSearchEmbeddings.ts:193]: **Unbounded IndexedDB growth.** `updatedAt` is written but never read — there's no TTL, size cap, or eviction, and `clearComponentSearchEmbeddingCache` is only called in tests. Every unique *query* is embedded and cached forever, so the table grows across sessions without limit (each 1536-dim vector is ~10KB+ as JSON). Add an eviction policy (TTL or max-rows prune), particularly for query embeddings. (The new `bulkPut` try/catch handles a full quota gracefully, but that's a backstop, not eviction.)

_(Moved inline from my stack review on #2432 — still open.)_

File: src/services/componentSearchEmbeddings.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+};
+
+componentSearchEmbeddingDb.version(1).stores({
+  embeddings: "cacheKey, embeddingModel, textHash",
+});
+
+interface EmbeddingOptions {
+  apiBase: string;
+  apiKey: string;
+  signal?: AbortSignal;
```
Comment (camielvs) [inline src/services/componentSearchEmbeddings.ts:42]: **Embedding fetches are uncancellable.** `EmbeddingOptions` has a `signal` field, but neither caller wires one up. Changing the query mid-flight leaves the billed embedding request running. This repo explicitly removed "uncancellable billed calls" for AI descriptions (see the `useComponentAiDescription` comment) — this reintroduces the pattern. Thread an `AbortSignal` tied to query change / unmount.

_(Moved inline from my stack review on #2432 — still open.)_

File: src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts [hunk 8]
```diff
@@ -47,11 +50,29 @@ import type {
 } from "@/utils/componentSpec";
 import { HOURS } from "@/utils/constants";
 
+function mergeUniqueMatches(
```
Comment (camielvs) [inline src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts:53]: **`mergeUniqueMatches` is copy-pasted verbatim** into both `useComponentSearchV2State` and `DashboardComponentsV2View`. And `buildComponentEmbeddingText` is now a *third* independent component→text projection alongside the index's `searchable` builder and `componentReferenceToCandidate` — note it doesn't apply the annotation-key exclusions or length caps the index uses, so the three can drift. Extract shared helpers; the embedding text in particular should probably reuse the index's extraction.

_(Moved inline from my stack review on #2432 — still open.)_

File: src/services/componentSearchEmbeddings.test.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+      index,
+      "open a spreadsheet",
+      OPTIONS,
+      { limit: 2 },
+    );
+
+    expect(results[0]?.digest).toBe("csv");
+  });
+
+  it("caches query and component embeddings in IndexedDB", async () => {
```
Comment (camielvs) [inline src/services/componentSearchEmbeddings.test.ts:200]: Good test coverage for the happy paths; consider adding one for a **>1-batch miss** (now that requests are chunked at 256) and for **cache invalidation when a component's text changes** (new hash → re-embed only that entry).

_(Moved inline from my stack review on #2432 — still open.)_

File: src/services/componentSearchEmbeddings.ts (hunk 6 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.ts:193]: > 🤖 This is an AI-generated response.

Fixed — the embedding cache now has both a TTL and max-entry pruning, and `updatedAt` is used for expiry/eviction.

File: src/services/componentSearchEmbeddings.ts (hunk 7 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.ts:42]: > 🤖 This is an AI-generated response.

Fixed — the editor embedding search now creates an AbortController, aborts previous in-flight embedding work on query changes/new searches/unmount, and passes the signal through to the embeddings fetch.

File: src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts (hunk 8 above)
Comment (Mbeaulne) [reply src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts:53]: > 🤖 This is an AI-generated response.

I agree the duplicate merge/projection helpers should be consolidated, but I kept that as a follow-up rather than expanding this PR into a broader dashboard/editor shared-search refactor.

File: src/services/componentSearchEmbeddings.test.ts (hunk 9 above)
Comment (Mbeaulne) [reply src/services/componentSearchEmbeddings.test.ts:200]: > 🤖 This is an AI-generated response.

Fixed — added coverage for multi-batch embedding misses and cache invalidation when a component's embedding text changes.

File: src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts [hunk 10]
```diff
@@ -218,18 +248,75 @@ export function useComponentSearchV2State(
...
+    },
   ) => {
     if (!trimmedQuery || matches.length === 0 || !isConfigured) return;
 
-    const rerankBase =
-      mode === "smart" ? matches.slice(0, LEXICAL_RESULT_LIMIT) : matches;
-    const candidates = rerankBase
+    const effectiveLimit =
+      mode === "smart" ? Math.min(limit, LEXICAL_RESULT_LIMIT) : limit;
+    const embeddingMatches =
```
Comment (maxy-shpfy) [inline src/routes/v2/pages/Editor/hooks/useComponentSearchV2State.ts]: > 🤖 This is an AI-generated code review comment.

The new embedding rerank path only runs after `startRerank` receives non-empty lexical candidates, and the returned `canRerank` flag is still based on `aiCandidateMatches.length > 0`. In the editor, a query with no lexical/source-diverse matches disables AI…

Suggested fix: Allow the embedding path to seed candidates from `index` when lexical candidates are empty (when an AI provider API base is configured), and update `canRerank`/the early return so semantic-only queries can trigger AI …

File: src/services/componentSearchEmbeddings.ts [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+  }
+
+  const embeddings = readEmbeddingResponse(payload, texts.length);
+  if (embeddings.length === 0) {
+    throw new Error("Embeddings API returned no usable embeddings");
+  }
+  return embeddings;
+}
+
+async function embedTextsWithCache(
```
Comment (maxy-shpfy) [inline src/services/componentSearchEmbeddings.ts:142]: > 🤖 This is an AI-generated code review comment.

The new service is 259 analyzed lines with cyclomatic complexity 43 and FTA score 61.27, both above the review thresholds. The cache lookup/fetch batching/persistence/pruning/ranking responsibilities are all in one module, which makes the failure modes hard…

Suggested fix: Split the IndexedDB cache adapter, embedding API client/response parsing, and ranking/scoring helpers into focused functions or sibling modules, then add targeted tests for each failure boundary.

**Review by maxy-shpfy — APPROVED** [review]

**Comment by Mbeaulne** [comment]
### Merge activity

* **Jun 26, 7:34 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2432).
* **Jun 26, 8:17 PM UTC**: [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2432) rebased this pull request as part of a merge.
* **Jun 26, 8:20 PM UTC**: @Mbeaulne merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2432).
